lxd/instance/drivers/driver_qemu.go

```diff
@@ -1559,115 +1559,242 @@ func (vm *qemu) generateQemuConfigFile(bus string, devConfs []*deviceConfig.RunC
 		return "", err
 	}
 
-	// Now add the dynamic parts of the config.
-	err = qemuSerial.Execute(sb, map[string]interface{}{
-		"architecture":     vm.architectureName,
-		"ringbufSizeBytes": qmp.RingbufSize,
-	})
+	err = vm.addCPUConfig(sb)
 	if err != nil {
 		return "", err
 	}
 
-	err = qemuSCSI.Execute(sb, map[string]interface{}{
-		"architecture": vm.architectureName,
-	})
+	err = vm.addMemoryConfig(sb)
 	if err != nil {
 		return "", err
 	}
 
-	err = qemuBalloon.Execute(sb, map[string]interface{}{
+	err = qemuDriveFirmware.Execute(sb, map[string]interface{}{
 		"architecture": vm.architectureName,
+		"roPath":       filepath.Join(vm.ovmfPath(), "OVMF_CODE.fd"),
+		"nvramPath":    vm.nvramPath(),
 	})
 	if err != nil {
 		return "", err
 	}
 
-	err = qemuRNG.Execute(sb, map[string]interface{}{
-		"architecture": vm.architectureName,
+	err = qemuControlSocket.Execute(sb, map[string]interface{}{
+		"path": vm.monitorPath(),
 	})
 	if err != nil {
 		return "", err
 	}
 
-	err = vm.addMemoryConfig(sb)
-	if err != nil {
-		return "", err
+	// getBusAddress is used to create any needed root ports and provide
+	// the bus and address that should be used by a device. It supports
+	// automatically setting up multi-function devices and optimize their use.
+	portNum := 0
+	devNum := 5
```

Comment:
@stgraber I've found that you can lower this to 1 and it still seems fine:

```
root@v1:~# lspci -PP
00:00.0 Host bridge: Intel Corporation 82G33/G31/P35/P31 Express DRAM Controller
00:01.0 PCI bridge: Red Hat, Inc. QEMU PCIe Root port
00:01.1 PCI bridge: Red Hat, Inc. QEMU PCIe Root port
00:01.2 PCI bridge: Red Hat, Inc. QEMU PCIe Root port
00:01.3 PCI bridge: Red Hat, Inc. QEMU PCIe Root port
00:01.4 PCI bridge: Red Hat, Inc. QEMU PCIe Root port
00:01.5 PCI bridge: Red Hat, Inc. QEMU PCIe Root port
00:1f.0 ISA bridge: Intel Corporation 82801IB (ICH9) LPC Interface Controller (rev 02)
00:1f.2 SATA controller: Intel Corporation 82801IR/IO/IH (ICH9R/DO/DH) 6 port SATA Controller [AHCI mode] (rev 02)
00:1f.3 SMBus: Intel Corporation 82801I (ICH9 Family) SMBus Controller (rev 02)
00:01.0/01:00.0 Unclassified device [00ff]: Red Hat, Inc. Virtio memory balloon (rev 01)
00:01.0/01:00.1 Unclassified device [00ff]: Red Hat, Inc. Virtio RNG (rev 01)
00:01.0/01:00.2 Communication controller: Red Hat, Inc. Device 1053 (rev 01)
00:01.0/01:00.3 Communication controller: Red Hat, Inc. Virtio console (rev 01)
00:01.1/02:00.0 SCSI storage controller: Red Hat, Inc. Virtio SCSI (rev 01)
00:01.2/03:00.0 Unclassified device [0002]: Red Hat, Inc. Virtio filesystem (rev 01)
00:01.3/04:00.0 VGA compatible controller: Red Hat, Inc. Virtio GPU (rev 01)
00:01.4/05:00.0 Ethernet controller: Red Hat, Inc. Virtio network device (rev 01)
00:01.5/06:00.0 Ethernet controller: Intel Corporation 82576 Virtual Function (rev 01)
```

Reply:
Indeed, I thought there were built-in devices before 05 but it was just our own stuff.

Reply:
Reduced to 1